### DISCUSSION TITLE
[v0.23rc] Implement `core.BaseRecordProxy` and scaffold

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#5702] [Show and tell]
I've been a 1+ user of PocketBase with a built Shopify app (React+PB) in production for 9+ months and been a semi regular contributor on the project with discussions and the odd oAuth [provider](https://github.com/pocketbase/pocketbase/pull/5588) 😅

When the [v0.23rc upgrade](https://pocketbase.io/v023upgrade/go/) dropped (outstanding work @ganigeorgiev) I was trawled to get into it, but what drove me to really jump into it and refactor my entire app's backend was when I discovered `core.BaseRecordProxy` going through the code, as expressed [here](https://github.com/pocketbase/pocketbase/discussions/5631#discussioncomment-10891022).

I'm sure this example implementation will clarify it further:

```go
// file: reward_record.go
package app

import (
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/types"
)

const (
	RewardsCollectionName = "rewards"
	RewardFieldAccount = "account" // accounts single relation
	RewardFieldCode      = "code"
	RewardFieldRedeemed = "redeemed"
)

type Reward struct {
	core.BaseRecordProxy
}

func NewReward(collection *core.Collection, accountId, code string) (*Reward, error) {
	if collection.Name != RewardsCollectionName {
		return nil, ErrInvalidCollection
	}

	record := core.NewRecord(collection)
	record.Set("id:autogenerate", "")

	reward := newReward(record)
	reward.Set(RewardFieldAccount, accountId)
	reward.Set(RewardFieldCode, code)

	return reward, nil
}

func newReward(record *core.Record) *Reward {
	reward := &Reward{}
	reward.SetProxyRecord(record)
	return reward
}

func (r Reward) Account() string {
	return r.GetString(RewardFieldAccount)
}

func (r Reward) Code() string {
	return r.GetString(RewardFieldCode)
}

func (r Reward) Redeemed() types.DateTime {
	return r.GetDateTime(RewardFieldRedeemed)
}

func (r *Reward) SetRedeemed(value types.DateTime) {
	r.Set(RewardFieldRedeemed, value)
}

```

Compounded with how easy it is to query:

```go
// file: reward_query.go
package app

import (
	"errors"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

var (
	ErrFindRewardsCollection = errors.New("find reward collection")
	ErrFindRewardsByAccount = errors.New("find rewards by account")
	ErrFindRewardById = errors.New("find reward by id")
)

func (a *App) FindRewardsCollection() (*core.Collection, error) {
	c, err := a.app.FindCachedCollectionByNameOrId(RewardsCollectionName)
	if err != nil {
		return nil, wrapErrNoRows(err, ErrFindRewardsCollection)
	}

	return c, nil
}

func (a *App) FindRewardById(accountId string, id string) (*Reward, error) {
	collection, err := a.FindRewardsCollection()
	if err != nil {
		return nil, wrapErr(err, ErrFindRewardById)
	}

	record, err := a.app.FindFirstRecordByFilter(
		collection,
		"account = {:account} && id = {:id}",
		dbx.Params{
			"account": accountId,
			"id":      id,
		},
	)

	if err != nil {
		return nil, wrapErrNoRows(err, ErrFindRewardById)
	}

	reward := newReward(record)
	return reward, nil
}

func (a *App) FindRewardsByAccount(accountId string, sort string, limit int, offset int) ([]*Reward, error) {
	collection, err := a.FindRewardsCollection()
	if err != nil {
		return nil, wrapErr(err, ErrFindRewardsByAccount)
	}

	records, err := a.app.FindRecordsByFilter(
		collection,
		"account = {:account}",
		sort,
		limit,
		offset,
		dbx.Params{
			"account": accountId,
		},
	)

	if err != nil {
		return nil, wrapErr(err, ErrFindRewardsByAccount)
	}

	result := []*Reward{}
	for _, record := range records {
		reward := newReward(record)
		result = append(result, reward)
	}
	return result, nil
}

func (a *App) RewardsQuery() *dbx.SelectQuery {
	return a.app.RecordQuery(RewardsCollectionName)
}
```

I think this deserves a `./pb scaffold collection <name>` command plugin, if there's interest in it comment below. I would really appreciate @ganigeorgiev views on this example of the `core.BaseRecordProxy`. 🙋‍♂️

**Comment by ganigeorgiev** [comment]
_Side-note: there are plans to introduce `go:generate` scripts or a custom command as you've suggested to autogenerate the record proxy types with all available collection fields (see also https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067), including dedicated type specific proxy hooks (e.g. `OnRewardCreate()`) but that will be after v0.23 (technically I already have a partial `go:generate` plugin implementation but I haven't published it since it needs more testing and I don't want to block further v0.23)._

**Reply by pnmcosta** [reply]
That is awesome, if you'd take it I'd volunteer to contribute with code for that command/generator, I've already started mine that I'm using for specific collections but I'd happily switch to help to contribute to unblock you from it.

Record specific proxy hooks would be awesome, I'm currently exposing direct ones from PB's app via my custom app, but that would be something even better.

**Comment by dschissler** [comment]
@pnmcosta I make my New model functions like this.  I noticed that you're passing the collection in whereas I'm passing in the core.App.  So I'm just wondering if after seeing this that you like it better, or if there is a reason that you're doing it the way that you are.  My thinking is that the way that you are doing it makes it rather verbose to create a new model by needing to error check for the collection, and again for the model.

I personally resent even having to error check for the empty model, but I know that it's necessary due to that collection possibly not existing.  It seems that you're doing all of that twice though.  So I'd just like to know if there are benefits to your approach that I'm not aware of.

```golang
func UsersCollection(app core.App) (*core.Collection, error) {
	return app.FindCollectionByNameOrId(Users)
}

func NewUser(app core.App) (*core.Record, error) {
	if col, err := UsersCollection(app); err != nil {
		return nil, err
	} else {
		return core.NewRecord(col), nil
	}
}
```

**Reply by pnmcosta** [reply]
That's one way, I purposely shadow PB with my own custom `app.App` struct that includes only the methods I shadow of PB or my own.

In my case it made more sense to depend on collection directly, but I do agree with:

> I personally resent even having to error check for the empty model

So much so that I have so many `reward, _ := app.NewReward(...)` that I'm considering panicking instead 😅
